### PR TITLE
Adds a new rule for disabling route resource partials

### PR DESCRIPTION
**Pull request by lukeraymonddowning**
This rule will prevent the use of partial route resources, such as:

```php
Route::resource('foo', 'FooController')->except(['show']);

Route::apiResource('foo', 'FooController')->only(['show']);
```

It will still allow route resources that make use of all the methods, and will also allow use of other methods such as `shallow`,  `names`, `parameters` and `scoped`. It also can detect chained route methods and still pick out `except` and `only`.

```php
Route::resource('foo', 'FooController')->scoped()->except(['show']);
```

### CODE REVIEW

**Review by olivernybroe — COMMENTED** [review]
Looks really nice 👍 

There are however two scenarios that I can think of which this does not catch.

```php
Route::group([], function(Router $router) {
	$router->resource('lukeness')->except('destroy');
}
```

And
```php
$router->resource('lukeness')->except('destroy');

```
because the `RouteFileRegistrar` gives a `$router` variable to the route file

File: src/PHPStan/Laravel/DisallowPartialRouteResourceRule.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+use Illuminate\Support\Facades\Route;
+use PhpParser\Node;
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleError;
+use PHPStan\Rules\RuleErrorBuilder;
+
+/**
+ * @implements Rule<FuncCall>
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply src/PHPStan/Laravel/DisallowPartialRouteResourceRule.php]: Lol, good catch. Updated

File: tests/PHPStan/Laravel/DisallowPartialRouteResourceRule/DisallowPartialRouteResourceRuleTest.php [hunk 2]
```diff
@@ -39,3 +41,24 @@
...
+        ],
+        [
+            'Usage of [only] method on route resource is disallowed. Please split the resource into multiple routes.',
+            7
+        ]
+    ],
+    'calls complete route resource' => [
+        __DIR__ . '/Fixture/route_resource.php.inc',
+    ],
+]);
```
Comment (olivernybroe) [inline tests/PHPStan/Laravel/DisallowPartialRouteResourceRule/DisallowPartialRouteResourceRuleTest.php:64]: newline missing ;)

File: src/PHPStan/Laravel/DisallowPartialRouteResource/PartialRouteResourceInspector.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        return false;
+    }
+
+    private function errorFor(string $method): RuleError
+    {
+        return RuleErrorBuilder::message(
+            "Usage of [{$method}] method on route resource is disallowed. Please split the resource into multiple routes."
+        )->build();
+    }
+}
```
Comment (olivernybroe) [inline src/PHPStan/Laravel/DisallowPartialRouteResource/PartialRouteResourceInspector.php:105]: newline

File: tests/PHPStan/Laravel/DisallowPartialRouteResourceRule/Fixture/grouped_route_resource_with_except.php.inc [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+use Illuminate\Support\Facades\Route;
+
+Route::group([], function ($router) {
+    $router->resource('foo', 'FooController')->except(['show']);
+    $router->resource('bar', 'BarController')->only(['show']);
+});
+
+// Some other method call to make sure everything still works
+$router->resource('baz', 'BazController');
```
Comment (olivernybroe) [inline tests/PHPStan/Laravel/DisallowPartialRouteResourceRule/Fixture/grouped_route_resource_with_except.php.inc:11]: missing newline

**Review by olivernybroe — APPROVED** [review]
Looks pretty good 👍 

Nice job

**Pull request merged**